### PR TITLE
[SPARK-23599][SQL] Use RandomUUIDGenerator in Uuid expression

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

As stated in Jira, there are problems with current `Uuid` expression which uses `java.util.UUID.randomUUID` for UUID generation.

This patch uses the newly added `RandomUUIDGenerator` for UUID generation. So we can make `Uuid` deterministic between retries.

## How was this patch tested?

Added unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -98,6 +99,8 @@ class Analyzer(
     this(catalog, conf, conf.optimizerMaxIterations)
   }
 
+  private lazy val random = new Random()
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Shall we put `random` in the `ResolvedUuidExpressions`? That makes it a little bit easier to follow.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolvedUuidExpressionsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  private lazy val uuid3 = Uuid().as('_uuid3)
+  private lazy val uuid1Ref = uuid1.toAttribute
+
+  private val analyzer = getAnalyzer(caseSensitive = true)
+
+  private def getUuidExpressions(plan: LogicalPlan): Seq[Uuid] = {
+    val uuids = new ArrayBuffer[Uuid]()
+    plan.transformUp {
+      case p =>
+        p.transformExpressionsUp {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolvedUuidExpressionsSuite.scala]: NIT: use `collect`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolvedUuidExpressionsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  private lazy val uuid1 = Uuid().as('_uuid1)
+  private lazy val uuid2 = Uuid().as('_uuid2)
+  private lazy val uuid3 = Uuid().as('_uuid3)
+  private lazy val uuid1Ref = uuid1.toAttribute
+
+  private val analyzer = getAnalyzer(caseSensitive = true)
+
+  private def getUuidExpressions(plan: LogicalPlan): Seq[Uuid] = {
+    val uuids = new ArrayBuffer[Uuid]()
+    plan.transformUp {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolvedUuidExpressionsSuite.scala]: Nit use `flatMap`?

**Comment by viirya** [comment]
@hvanhovell Thanks! Your comments are addressed.

**Review by hvanhovell — APPROVED** [review]
LGTM - merging to master. Thanks!

**Comment by hvanhovell** [comment]
@viirya can you create a backport for 2.3?

**Comment by viirya** [comment]
@hvanhovell Ok. But this needs #20817. Since #20817 just adds new class and doesn't change existing code, I think it can be directly merged into 2.3. Should I create a backport PR of it too? Or you can direct backport it?

**Comment by hvanhovell** [comment]
@viirya I have backported #20817 to 2.3

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -1994,6 +1996,20 @@ class Analyzer(
...
+  /**
+   * Set the seed for random number generation in Uuid expressions.
+   */
+  object ResolvedUuidExpressions extends Rule[LogicalPlan] {
+    private lazy val random = new Random()
+
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.transformUp {
+      case p if p.resolved => p
+      case p => p transformExpressionsUp {
+        case Uuid(None) => Uuid(Some(random.nextLong()))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: shall we do the same thing for `Rand`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: hmm, if we want to make it deterministic between re-tries of same query. I think we should do it. I can make a PR for it, WDYT?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: SGTM. We can even create a base trait for these random functions.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I have checked, and actually `Rand` and `Randn` already have this behavior that they are deterministic between re-tries, though their random seeds are not determined at analysis but at constructing.

So I'm thinking should we do the same thing (random seed initialized at analysis) to `Rand` and `Randn`? Besides just to be consistent with `Uuid`, is any good reason to do this?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: One thing I think is, is any use case that we need to re-initialize random seed for `Rand`? Maybe streaming? For streaming query, I think `Rand` should use different random seed in each execution. For now, the random seed is initialized when constructing, even we re-analyze the query, it still uses same seed.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: what's the current behavior for rand in streaming?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I created a PR #21854 which shows behavior of `Uuid` in streaming. I think rand should be the same.

**Pull request closed without merging**